proto/kv-connect.md

```diff
@@ -0,0 +1,448 @@
+the resolved URL of the selected endpoint. The request body MUST be a Protobuf
+message in the format `com.deno.kv.datapath.SnapshotRead` (defined in the
+`schema/datapath.proto` file in this repository).
+
+<!-- If the client wants to receive a stream of updates from the server, rather than
+just a single response, the client MUST include a `x-denokv-stream` header with
+the value `true`. -->
+
+Upon receiving the POST request, the server MUST verify that the request body
+adheres to the Protobuf schema for the `com.deno.kv.datapath.SnapshotRead`
+message. If the request body is invalid, the server MUST respond with a 4xx
+response and a plain text body containing a human readable error message.
+
+The server MAY return a 5xx class HTTP status response at any time if the server
+is unable to process the request right now due to a temporary condition. The
+server SHOULD include a plain text body with a human readable error message in
+this case.
+
+The server MAY perform quota checks on the request. If the request is rejected
+due to quota limits, the server MUST respond with a 4xx class response and a
+plain text body containing a human readable error message.
+
+<!-- If the `x-denokv-stream` header is included and has the value `true`, the server
+MUST perform the requested snapshot read operation. If the operation fails due
+to a malformed request, the server MUST respond with a 4xx class response
+containing a plain text body with a human readable error message. If the
+operation succeeds the server must respond with a 200 OK response. The response
+MUST include a `Content-Type` header with the value `...`. The response body
+MUST be a stream of Protobuf -->
+
+The server MUST perform the requested snapshot read operation. The the server
+MUST respond with a 200 OK response. The response MUST include a `Content-Type`
+header with the value `application/x-protobuf`. The response body MUST be a
+Protobuf message in the format `com.deno.kv.datapath.SnapshotReadOutput`. If the
+server is unable to perform the read because the database is not available from
+this server, the `read_disabled` field MUST be set to `true`. The server MUST
+include the `read_is_strongly_consistent` field, setting it to `true` if the
+server has a strongly consistent view of the database, and `false` if the server
+has an eventually consistent view. If the request succeeds, the server MUST
+include the `ranges` field with a list of ranges that were read from the
+database. The order MUST be the same as the order of the ranges in the request.
+
+If the client fails to receive a response from the server due to a network
+error, or a 5xx class HTTP status, the client SHOULD retry the request using an
+exponential backoff strategy with unlimited retries.
+
+The client MUST verify that the response is a 200 OK response. If the response
+is a 4xx class error response, the client MUST fatally abort the request and
```

Comment:
Similar comment to above about rate limiting. Should the server be able to respond with 429? and should the client interpret that as retry with backoff?

Reply:
I think the server should respond with a 503 then. It's easier if 4xx == never retry, and 5xx == retry with backoff.

Reply:
I think there should be a mechanism to "gracefully" rate limit, like starting to throttle at 90% of the limit and when it reaches 100% wait until the end of the rate limit window